File: android/app/src/main/java/com/kryptos/android/core/PasswordCipher.kt
package com.kryptos.android.core

import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

object PasswordCipher {
    const val SALT_LEN = 16
    const val TAG_LEN = 16
    const val KEY_LEN = 32
    const val DERIVED_LEN = 44

    private const val TAG_BITS = 128

    fun sealBody(plaintext: ByteArray, key: ByteArray, nonce: ByteArray, version: Byte, pad: Boolean): ByteArray {
        val compressed = Deflate.compress(plaintext)
        val deflate = compressed != null
        val content = if (deflate) compressed!! else plaintext
        val framed = if (pad) Padding.frame(content) else content
        val body = ByteArray(1 + framed.size)
        body[0] = ((if (deflate) 0x01 else 0x00) or (if (pad) 0x02 else 0x00)).toByte()
        framed.copyInto(body, 1)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(byteArrayOf(version))
        return cipher.doFinal(body)
    }

    fun openBody(sealed: ByteArray, key: ByteArray, nonce: ByteArray, version: Byte): ByteArray {
        if (sealed.size < TAG_LEN) throw CipherException(CipherException.Kind.DECRYPTION_FAILED)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(byteArrayOf(version))
        val body = try {
            cipher.doFinal(sealed)
        } catch (e: Exception) {
            throw CipherException(CipherException.Kind.DECRYPTION_FAILED)
        }
        if (body.isEmpty()) throw CipherException(CipherException.Kind.DECRYPTION_FAILED)
        val flag = body[0].toInt()
        var content = body.copyOfRange(1, body.size)
        if (flag and 0x02 != 0) {
            content = Padding.unframe(content) ?: throw CipherException(CipherException.Kind.DECRYPTION_FAILED)
        }
        if (flag and 0x01 != 0) {
            return Deflate.decompress(content) ?: throw CipherException(CipherException.Kind.DECRYPTION_FAILED)
        }
        return content
    }

    fun encrypt(plaintext: ByteArray, password: String, pad: Boolean = false): ByteArray {
        val salt = randomBytes(SALT_LEN)
        val version = Argon2id.PROFILE_VERSION
        val km = Argon2id.derive(password, salt, DERIVED_LEN)
        val key = km.copyOfRange(0, KEY_LEN)
        val nonce = km.copyOfRange(KEY_LEN, DERIVED_LEN)
        km.fill(0)
        val sealed = try {
            sealBody(plaintext, key, nonce, version, pad)
        } finally {
            key.fill(0)
            nonce.fill(0)
        }
        return salt + version + sealed
    }

    fun decrypt(data: ByteArray, password: String): ByteArray {
        if (data.size < SALT_LEN + 1 + TAG_LEN) throw CipherException(CipherException.Kind.MALFORMED)
        val salt = data.copyOfRange(0, SALT_LEN)
        val version = data[SALT_LEN]
        if (version != Argon2id.PROFILE_VERSION) throw CipherException(CipherException.Kind.MALFORMED)
        val sealed = data.copyOfRange(SALT_LEN + 1, data.size)
        val km = Argon2id.derive(password, salt, DERIVED_LEN)
        val key = km.copyOfRange(0, KEY_LEN)
        val nonce = km.copyOfRange(KEY_LEN, DERIVED_LEN)
        km.fill(0)
        try {
            return openBody(sealed, key, nonce, version)
        } finally {
            key.fill(0)
            nonce.fill(0)
        }
    }
}

File: CipherCore/Sources/CipherCore/Symmetric/PasswordCipher.swift
import Foundation
import CryptoKit

public enum PasswordCipher {
    public static let saltLength = 16
    static let tagLength = 16
    static let keyLength = 32
    static let nonceLength = 12
    static let derivedLength = 44

    static func split(_ derived: [UInt8]) throws -> (SymmetricKey, AES.GCM.Nonce) {
        guard derived.count >= keyLength + nonceLength else { throw CipherError.invalidInput }
        let key = SymmetricKey(data: Data(derived[0 ..< keyLength]))
        let nonce = try AES.GCM.Nonce(data: Data(derived[keyLength ..< keyLength + nonceLength]))
        return (key, nonce)
    }

    static func sealBody(_ plaintext: Data, key: SymmetricKey, nonce: AES.GCM.Nonce,
                         version: UInt8, pad: Bool) throws -> Data {
        let compressed = Deflate.compress(plaintext)
        let deflate = compressed != nil
        let content = deflate ? compressed! : plaintext
        let framed = pad ? Padding.frame(content) : content
        var body = Data([(deflate ? 0x01 : 0x00) | (pad ? 0x02 : 0x00)])
        body.append(framed)
        let sealed = try AES.GCM.seal(body, using: key, nonce: nonce, authenticating: Data([version]))
        return sealed.ciphertext + sealed.tag
    }

    static func openBody(_ sealed: Data, key: SymmetricKey, nonce: AES.GCM.Nonce,
                         version: UInt8) throws -> Data {
        guard sealed.count >= tagLength else { throw CipherError.decryptionFailed }
        let ct = sealed.prefix(sealed.count - tagLength)
        let tag = sealed.suffix(tagLength)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: Data(ct), tag: Data(tag))
        let body: Data
        do {
            body = try AES.GCM.open(box, using: key, authenticating: Data([version]))
        } catch {
            throw CipherError.decryptionFailed
        }
        guard let flag = body.first else { throw CipherError.decryptionFailed }
        var content = Data(body.suffix(from: body.startIndex + 1))
        if flag & 0x02 != 0 {
            guard let unframed = Padding.unframe(content) else { throw CipherError.decryptionFailed }
            content = unframed
        }
        if flag & 0x01 != 0 {
            guard let inflated = Deflate.decompress(content) else { throw CipherError.decryptionFailed }
            return inflated
        }
        return content
    }

    public static func encrypt(_ plaintext: Data, password: String, pad: Bool = false) throws -> Data {
        let salt = randomBytes(saltLength)
        let version = Argon2id.profileVersion
        var derived = try Argon2id.derive(password: Data(password.utf8), salt: salt, length: derivedLength)
        defer { Argon2id.zero(&derived) }
        let (key, nonce) = try split(derived)
        var out = salt
        out.append(version)
        out.append(try sealBody(plaintext, key: key, nonce: nonce, version: version, pad: pad))
        return out
    }

    public static func decrypt(_ data: Data, password: String) throws -> Data {
        guard data.count >= saltLength + 1 + tagLength else { throw CipherError.malformed }
        let salt = Data(data.prefix(saltLength))
        let version = data[data.startIndex + saltLength]
        guard version == Argon2id.profileVersion else { throw CipherError.malformed }
        let sealed = Data(data.suffix(from: data.startIndex + saltLength + 1))
        var derived = try Argon2id.derive(password: Data(password.utf8), salt: salt, length: derivedLength)
        defer { Argon2id.zero(&derived) }
        let (key, nonce) = try split(derived)
        return try openBody(sealed, key: key, nonce: nonce, version: version)
    }
}
